polybuild/polybuild.py

```diff
@@ -0,0 +1,358 @@
+        if self.meta.is_cxx:
+            compile_command.append("gclang++")
+        else:
+            compile_command.append("gclang")
+        compile_command += ["-pie", "-fPIC"]
+        if self.meta.is_cxx:
+            compile_command.append("-stdlib=libc++")
+            compile_command.append("-nostdinc++")
+            compile_command.append("-I" + self.meta.compiler_dir + "/../cxx_libs/include/c++/v1/")
+            compile_command.append("-L" + self.meta.compiler_dir + "/../cxx_libs/lib/")
+        for arg in argv[1:]:
+            compile_command.append(arg)
+        is_linking = self.poly_is_linking(argv)
+        if is_linking:
+            # If its cxx, link in our c++ libs
+            if self.meta.is_cxx:
+                compile_command += ["-lc++", "-lc++abipoly", "-lc++abi", "-lpthread"]
+
+        command = shlex.split(" ".join(shlex.quote(arg) for arg in compile_command))
+        res = subprocess.call(command)
+        if res != 0:
+            return False
+        return True
+
+
+"""
+Store a build artifact to the artifact storage via copy
+"""
+
+
+def store_artifact(file_path, artifact_path) -> bool:
+    filename = get_file_name(file_path)
+    artifact_file_path = artifact_path + "/" + filename
+    # Check if its an absolute path
+    if file_path[0] != "/":
+        cwd = os.getcwd()
+        targ_path = cwd + "/" + file_path
+    else:
+        targ_path = file_path
+    if not os.path.exists(targ_path):
+        print(f"Error! cannot find {targ_path}")
+        return False
+    ret = subprocess.call(["cp", targ_path, artifact_file_path])
+    if ret != 0:
+        print(f"Error! failed to store {targ_path}, error was {ret}")
+        print(f"Artifact path: {artifact_file_path}")
+        return False
+    return True
+
+
+def get_file_name(file_path: str) -> str:
+    last_slash: int = file_path.rfind("/")
+    # If not path just return the file
+    if last_slash == -1:
+        return file_path
+    # Return the file name from last path
+    return file_path[last_slash + 1 :]
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="""
+    Compiler wrapper around gllvm and instrumentation driver for PolyTracker
+
+    For programs with a simple build system, you can quickstart the process
+    by invoking:  
+
+    Compile normally by invoking polybuild <whatever your arguments are> 
+    These arguments will get passed to gclang/gclang++
+
+    Extract the bitcode from the built target
+    get-bc -b <target_binary> 
+
+    OPTIONAL: Link multiple bitcode files together with llvm-link
+
+    Instrument that whole program bitcode file by invoking 
+    polybuild --instrument-bitcode -f <bitcode_file.bc> -o <output_file> 
+    """
+    )
+    parser.add_argument("--instrument-bitcode", action="store_true", help="Specify to add polytracker instrumentation")
+    parser.add_argument("--input-file", "-f", type=str, default=None, help="Path to the whole program bitcode file")
```

Comment:
If you take out `default=None,` then `argparse` will automatically fail with a nice error message if it is omitted and you can remove your manual check below.

Reply:
can you explain this further? So when I actually decide to parse the args, if there is no default value, it just fails and says i need it? that sounds nice, ill do that.

Reply:
Done!